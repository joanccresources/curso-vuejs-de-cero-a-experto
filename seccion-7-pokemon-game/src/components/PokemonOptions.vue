<template>
  <div class="mt-4">
    <ul class="list-group">
      <!-- Debemos ponerle su :key -->
      <li
        v-for="pokemon in pokemons"
        :key="pokemon.id"
        class="list-group-item list-group-item-action"
        @click="$emit('selectionPokemon', pokemon.id)"
      >
        {{ pokemon.name }}
      </li>
    </ul>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";

interface Pokemon {
  id: number;
  name: string;
}

export default defineComponent({
  props: {
    pokemons: {
      type: Array as () => Array<Pokemon>,
      required: true,
    },
  },
  data() {
    return {
      // pokemons=this.pokemons
    };
  },

  // Despues de renderizar se ejecuta el mounted
  // Es parecido al useEffect
  // Ciclo de vida del componente
  mounted() {
    console.log("mounted 2");
  },
});
</script>

<style scoped>
.list-group-item {
  cursor: pointer;
}
</style>
