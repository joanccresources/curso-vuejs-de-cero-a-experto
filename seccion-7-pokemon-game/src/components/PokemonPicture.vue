<template>
  <div class="pokemon-container">
    <img :src="imgSrc" alt="pokemon" class="img-fluid img-hidden" />
    <img
      v-if="showPokemon"
      :src="imgSrc"
      alt="pokemon"
      class="img-fluid fade-in"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";

export default defineComponent({
  props: {
    pokemonId: {
      type: Number,
      required: true,
    },
    showPokemon: {
      type: Boolean,
      required: true,
      default: false,
    },
  },

  data(): { id: number } {
    return {
      id: this.pokemonId,
    };
  },

  computed: {
    imgSrc(): string {
      return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.id}.svg`;
    },
  },
});
</script>

<style lang="scss" scoped>
.pokemon-container {
  height: 200px;
}
img {
  height: 200px;
  position: absolute;
  right: 32%;
  user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-drag: none;
  -webkit-user-select: none;
}
.img-hidden {
  filter: brightness(0);
}
</style>
