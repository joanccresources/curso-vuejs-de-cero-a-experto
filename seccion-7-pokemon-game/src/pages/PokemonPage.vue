<template>
  <div class="container">
    <h1 v-if="!pokemon">Espere porfavor...</h1>
    <div v-else>
      <h1>¿Quién es este pokémon?</h1>
      <!--  -->
      <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />
      <PokemonOptions :pokemons="pokemonArr" @selection-pokemon="checkAnswer" />

      <div v-if="showAnswer" class="fade-in">
        <div
          class="alert h3 mt-3"
          :class="{ 'alert-success': isCorrect, 'alert-danger': !isCorrect }"
        >
          {{ message }}
        </div>
        <button @click="newGame" class="btn btn-primary mt-1">
          Nuevo juego!
        </button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
interface Pokemon {
  id: number;
  name: string;
}
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import { getPokemonOptions } from "@/helpers";

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArr: [] as Pokemon[],
      // Le ponemos null para que solo cuando se un objeto mande la imagen
      pokemon: null as Pokemon | null,
      showPokemon: false as boolean,
      // Para decirle si esta bien o mal lo seleccionado
      showAnswer: false as boolean,
      message: "" as string,
      isCorrect: true as boolean,
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      // Esto imprime un objeto y la data esta en [[Target]]
      // console.log(this.pokemonArr);

      // Debemos generar un aleatorio [0,3]
      const rndInt: number = Math.floor(Math.random() * 4);

      // Y ahora obtenemos el pokemon de esa posicion
      this.pokemon = this.pokemonArr[rndInt];
    },

    checkAnswer(pokemonSelectedId: number) {
      this.showPokemon = true;
      // Cuando hagamos click recien se vera el mensaje
      this.showAnswer = true;

      // Si son iguales es porque es el correcto
      if (pokemonSelectedId === this.pokemon!.id) {
        this.message = `Correcto, ${this.pokemon!.name}`;
        this.isCorrect = true;
      } else {
        this.message = `Oops, era ${this.pokemon!.name}`;
        this.isCorrect = false;
      }
    },

    newGame() {
      // Reiniciamos todas las variables de data
      this.pokemonArr = [];
      this.pokemon = null;
      this.showPokemon = false;
      this.showAnswer = false;
      this.message = "";
      this.isCorrect = true;

      this.mixPokemonArray();
    },
  },

  // Despues de renderizar se ejecuta el mounted
  // Es parecido al useEffect
  // Ciclo de vida del componente
  mounted() {
    console.log("mounted 1");
    // Una vez montado ejecutamos el metodo
    this.mixPokemonArray();
  },
};
</script>
