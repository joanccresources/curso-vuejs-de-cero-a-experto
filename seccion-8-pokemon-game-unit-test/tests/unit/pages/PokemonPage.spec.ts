import { ComponentPublicInstance } from "vue";
import { shallowMount, mount, VueWrapper } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage.vue";
import { mockPokemons } from "../mocks";
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
interface Pokemon {
  id: number;
  name: string;
}

describe("PokemonPage Component", () => {
  let wrapper: VueWrapper<any, ComponentPublicInstance<{}, any>>;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);

    jest.clearAllMocks();
  });

  test("debe hacer match con el snapshoot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Debe de llamar mixPokemonArray al montar", () => {
    // ---Metodo de la clase
    // const mixPokemonArraySpy = jest.spyOn(
    //   PokemonPage.methods!,
    //   "mixPokemonArray"
    // );
    // expect(mixPokemonArraySpy).toHaveBeenCalled();
    // ---Este metodo no funciono y si funciono en otro
    // const mixPokemonArraySpy = jest.spyOn(wrapper.vm, "mixPokemonArray");
    // expect(mixPokemonArraySpy).toHaveBeenCalled();
    // Expected number of calls: 1
    // Received number of calls: 0
    // ---ChatGpt
    // const mixPokemonArrayMock = jest.fn();
    // const wrapper = shallowMount(PokemonPage, {
    //   methods: {
    //     mixPokemonArray: mixPokemonArrayMock,
    //   },
    // });
  });

  test("debe hacer match con el snapshot cuando cargan los pokemons", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: mockPokemons as Pokemon[],
          pokemon: mockPokemons[0] as Pokemon | null,
          showPokemon: false as boolean,
          showAnswer: false as boolean,
          message: "" as string,
          isCorrect: true as boolean,
        };
      },
    });
    const wrapper2 = mount(PokemonPage, {
      data() {
        return {
          pokemonArr: mockPokemons as Pokemon[],
          pokemon: mockPokemons[0] as Pokemon | null,
          showPokemon: false as boolean,
          showAnswer: false as boolean,
          message: "" as string,
          isCorrect: true as boolean,
        };
      },
    });
    // ---- Te imprime un pedaso nomas algo ligero
    // console.log(wrapper.html());
    //   <div class="container">
    //   <div>
    //     <h1>¿Quién es este pokémon?</h1><!--  -->
    //     <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
    //     <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    //     <!--v-if-->
    //   </div>
    // </div>

    // ----Te imprime Todo
    // console.log(wrapper2.html());
    // <div class="container">
    //   <div>
    //     <h1>¿Quién es este pokémon?</h1><!--  -->
    //     <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/5.svg" alt="pokemon" class="img-fluid img-hidden"></div>
    //     <div class="mt-4">
    //       <ul class="list-group">
    //         <!-- Debemos ponerle su :key -->
    //         <li class="list-group-item list-group-item-action">bulbasaur</li>
    //         <li class="list-group-item list-group-item-action">ivysaur</li>
    //         <li class="list-group-item list-group-item-action">venusaur</li>
    //         <li class="list-group-item list-group-item-action">charmander</li>
    //       </ul>
    //     </div>
    //     <!--v-if-->
    //   </div>
    // </div>
    expect(wrapper.html()).toMatchSnapshot();
    // Siempre que sea posible usar el shallowMount
  });

  test("debe de mostrar los componentes de PokemonPicture y PokemonOptions", () => {
    // PokemonPicture debe de existir
    // PokemonOptions debe de existir
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: mockPokemons as Pokemon[],
          pokemon: mockPokemons[0] as Pokemon | null,
          showPokemon: false as boolean,
          showAnswer: false as boolean,
          message: "" as string,
          isCorrect: true as boolean,
        };
      },
    });
    const picture = wrapper.find("pokemon-picture-stub");
    const options = wrapper.find("pokemon-options-stub");
    expect(picture.exists()).toBeTruthy();
    expect(options.exists()).toBeTruthy();

    // PokemonPicture attribute pokemonid===5 (5 es el codigo del primero pokemon)
    // PokemonOptions attribute pokemons toBe true

    expect(picture.attributes("pokemonid")).toBe(mockPokemons[0].id.toString());
    expect(options.attributes("pokemons")).toBeTruthy();
  });

  test("pruebas con checkAnswer", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: mockPokemons as Pokemon[],
          pokemon: mockPokemons[0] as Pokemon | null,
          showPokemon: false as boolean,
          showAnswer: false as boolean,
          message: "" as string,
          isCorrect: true as boolean,
        };
      },
    });

    // Ejecuta el metodo literalmente
    // Le mandamos el id del primer pokemon
    await wrapper.vm.checkAnswer(5);

    // Con esto validamos que cuando se ejecuta la funcion "checkAnswer"
    // entonces ".title-h2" existe
    expect(wrapper.find(".title-h2").exists()).toBeTruthy();

    // console.log(wrapper.find(".title-h2").text()); // Correcto, bulbasaur

    // Si entro a checkAnswer entonces showPokemon debe ser true
    expect(wrapper.vm.showPokemon).toBeTruthy();

    expect(wrapper.find(".title-h2").text()).toBe(
      `Correcto, ${mockPokemons[0].name}`
    );

    // Ejecuta el metodo literalmente
    // le mandamos el id del segundo pokemon
    await wrapper.vm.checkAnswer(10);
    // console.log(wrapper.vm.message);
    expect(wrapper.vm.message).toBe(`Oops, era ${mockPokemons[0].name}`);
  });
});
