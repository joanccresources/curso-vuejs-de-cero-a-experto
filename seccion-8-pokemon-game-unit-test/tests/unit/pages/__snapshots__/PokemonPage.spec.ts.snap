// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe hacer match con el snapshoot 1`] = `
<div class="container">
  <h1>Espere porfavor...</h1>
</div>
`;

exports[`PokemonPage Component debe hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div class="container">
  <div>
    <h1>¿Quién es este pokémon?</h1><!--  -->
    <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
`;
